### PR TITLE
MGLStyleLayer predicates (filters)

### PR DESCRIPTION
**Pull request by frederoni**
wip #5973.

/cc @incanus @1ec5

### CODE REVIEW

File: platform/darwin/src/MGLBaseStyleLayer.h [hunk 1]
```diff
@@ -14,4 +14,11 @@
  */
 @property (nonatomic, assign) float minimumZoomLevel;
 
+@property (nonatomic) NSPredicate *predicate;
+
+/**
+ Updates the layer’s layout and paint properties.
+ */
+- (void)update;
```
Comment (1ec5) [inline platform/darwin/src/MGLBaseStyleLayer.h]: We made this method private in #6018. Any reason you’re reintroducing this method in the public interface?

**Comment by 1ec5** [comment]
To convert from NSPredicate to the various `mbgl::style::Filter` classes, we should use a visitor pattern like [the ones in MGLFeature.mm](https://github.com/mapbox/mapbox-gl-native/blob/bca9d091805dc01a4456ab3f24e9de87f9b4aa48/platform/darwin/src/MGLFeature.mm#L119). Whereas `PropertyValueEvaluator` converts a recursive (arbitrarily nested) C++ structure into an Objective-C structure, we want to convert a recursive Objective-C structure into a C++ structure. We can nonetheless rely on operator overloading in C++ to keep things clean without a series of if/else-if statements.

File: platform/darwin/src/MGLBaseStyleLayer.h (hunk 1 above)
Comment (frederoni) [inline platform/darwin/src/MGLBaseStyleLayer.h]: No reason. It was in my git stash from last week. I'll remove it.

File: platform/darwin/src/NSPredicate+MGLAdditions.mm [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#import "NSPredicate+MGLAdditions.h"
+
+class FilterEvaluator {
+public:
+    
+    mbgl::Value getValue(id obj) {
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.mm]: Instead of a catch-all `getValue()` method, use the Visitor pattern that we’re already inside. Define operators for NSString and NSNumber in a FilterExpressionEvaluator, then apply the FilterExpressionEvaluator to each `constantValue` you encounter.

File: platform/darwin/src/NSPredicate+MGLAdditions.mm [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        if ([obj isKindOfClass:NSString.class]) {
+            NSString *string = (NSString *)obj;
+            return std::string(string.UTF8String);
+        } else if ([obj isKindOfClass:NSString.class]) {
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.mm]: I think you meant NSNumber here.

File: platform/darwin/src/NSPredicate+MGLAdditions.mm [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                filter.value = getValue(predicate.rightExpression.constantValue);
+                return filter;
+            }
+            case NSNotEqualToPredicateOperatorType: {
+                auto filter = mbgl::style::NotEqualsFilter();
+                filter.key = predicate.leftExpression.keyPath.UTF8String;
+                filter.value = getValue(predicate.rightExpression.constantValue);
+                return filter;
+            }
+            default:
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.mm]: Instead of a `default` case that raises, add cases for all of the `NSPredicateOperatorType` values. If Foundation ever adds a new `NSPredicateOperatorType` value, we’ll get a compiler warning, but not if there’s a `default` case.

File: platform/darwin/src/NSPredicate+MGLAdditions.mm [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        return mbgl::style::NoneFilter();
+    }
+};
+
+@implementation NSPredicate (MGLAdditions)
+
+- (mbgl::style::Filter)mgl_filter
+{
+    if ([self isKindOfClass:NSCompoundPredicate.class]) {
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.mm]: Instead of this check, implement an operator in FilterEvaluator for NSCompoundPredicate and raise the “not yet supported” exception there.

File: platform/darwin/src/MGLCircleStyleLayer.h [hunk 6]
```diff
@@ -17,6 +17,8 @@ typedef NS_ENUM(NSUInteger, MGLCircleStyleLayerCirclePitchScale) {
 
 @interface MGLCircleStyleLayer : MGLBaseStyleLayer <MGLStyleLayer>
 
+@property (nonatomic) NSPredicate *predicate;
```
Comment (1ec5) [inline platform/darwin/src/MGLCircleStyleLayer.h]: #6076 has landed, so without a nullability qualifier, the compiler will assume that the predicate is non-`nil`.

**Comment by 1ec5** [comment]
Also make sure to test some edge cases where you may need to escape the format string or use format string specifiers:

* [ ] Keys with hyphens are converted correctly
* [ ] `$type` and `$id` keys (#5708) are converted correctly

File: platform/darwin/src/NSPredicate+MGLAdditions.mm [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        [NSException raise:@"Nested predicates (NSCompoundPredicate) are not yet supported" format:@""];
+    }
+    
+    NSComparisonPredicate *predicate = (NSComparisonPredicate *)self;
+    FilterEvaluator evaluator;
+    auto filter = evaluator(predicate);
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.mm]: A C++ evaluator may not be a great tool for converting from NSPredicate to `mbgl::style::Filter` after all. We generally use this pattern with `mapbox::util::variant`, not an Objective-C umbrella class. Instead, let’s make this method do nothing but assert, then implement `-mgl_filter` overrides in categories on NSComparisonPredicate and NSCompoundPredicate.

The logic for converting a key or value to an `mbgl::Value` should live in a category on NSExpression.

A C++ evaluator remains a good idea for efficiently and compactly converting an `mbgl::style::Filter` into an NSPredicate. We could have the evaluator call into category initializers on NSComparisonPredicate, NSCompoundPredicate, and NSExpression, if that would make the code easier to understand.

File: platform/darwin/src/MGLCircleStyleLayer.h (hunk 6 above)
Comment (frederoni) [inline platform/darwin/src/MGLCircleStyleLayer.h]: will fix that after rebase

File: platform/darwin/src/NSPredicate+MGLAdditions.h [hunk 8]
```diff
@@ -1,9 +1,108 @@
...
 @end
+
+namespace mbgl {
+namespace darwin {
+
+class FilterEvaluator {
+public:
+
+mbgl::style::Filter operator()(NSComparisonPredicate *predicate) {
+    switch (predicate.predicateOperatorType) {
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.h]: The entire FilterEvaluator definition shouldn't be inlined in the header. Declare the operators here but define them in the implementation file.

File: platform/darwin/src/NSComparisonPredicate+MGLAdditions.h [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#import <Foundation/Foundation.h>
+
+#include <mbgl/style/filter.hpp>
+
+@interface NSComparisonPredicate (MGLAdditions)
+
+- (mbgl::style::Filter)mgl_filter;
```
Comment (1ec5) [inline platform/darwin/src/NSComparisonPredicate+MGLAdditions.h]: Since this is already declared on NSPredicate(MGLAdditions), there's no need to declare it here and in NSCompoundPredicate(MGLAdditions).

File: platform/darwin/src/NSPredicate+MGLAdditions.h [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+@interface NSPredicate (MGLAdditions)
+
+- (mbgl::style::Filter)mgl_filter;
+
+@end
+
+namespace mbgl {
+namespace darwin {
+
+class FilterEvaluator {
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.h]: This class is no longer needed. Fold the NSComparisonPredicate operator into `-[NSComparisonPredicate(MGLAdditions) mgl_filter]` and the NSComound Predicate operator into `-[NSCompoundPredicate(MGLAdditions) mgl_filter]`.

File: platform/darwin/src/NSPredicate+MGLAdditions.mm [hunk 11]
```diff
@@ -0,0 +1,11 @@
+#import "NSPredicate+MGLAdditions.h"
+
+@implementation NSPredicate (MGLAdditions)
+
+- (mbgl::style::Filter)mgl_filter
+{
+    [NSException raise:@"A specific predicate must override this method." format:@""];
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.mm]: The raise parameter should be a short title. The format parameter should be a complete sentence describing the problem.

File: platform/darwin/src/NSExpression+MGLAdditions.mm [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    [NSException raise:@"Values not handled" format:@""];
+    return std::vector<mbgl::Value>();
+}
+
+- (mbgl::Value)mgl_filterValue
+{
+    return [self convertedValueWithValue:self.constantValue];
+}
+
+- (mbgl::Value)convertedValueWithValue:(id)value
```
Comment (1ec5) [inline platform/darwin/src/NSExpression+MGLAdditions.mm]: Always prefix with mgl_ when adding a category method on a class we don't own.

File: platform/darwin/src/NSPredicate+MGLAdditions.h [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+mbgl::style::Filter operator()(NSComparisonPredicate *predicate) {
+    switch (predicate.predicateOperatorType) {
+        case NSMatchesPredicateOperatorType:
+        case NSLikePredicateOperatorType:
+        case NSBeginsWithPredicateOperatorType:
+        case NSEndsWithPredicateOperatorType:
+        case NSCustomSelectorPredicateOperatorType:
+        case NSContainsPredicateOperatorType:
+        case NSBetweenPredicateOperatorType:
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.h]: These are all operators we don't support. They should not be conflated with the equals operator. Instead of falling through, raise an exception explaining that the operator is unsupported.

Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.h]: As it happens, we can implement between as AllFilter(GreaterThanEqualsFilter, LessThanEqualsFilter).

File: platform/darwin/src/NSPredicate+MGLAdditions.h [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+public:
+
+mbgl::style::Filter operator()(NSComparisonPredicate *predicate) {
+    switch (predicate.predicateOperatorType) {
+        case NSMatchesPredicateOperatorType:
+        case NSLikePredicateOperatorType:
+        case NSBeginsWithPredicateOperatorType:
+        case NSEndsWithPredicateOperatorType:
+        case NSCustomSelectorPredicateOperatorType:
+        case NSContainsPredicateOperatorType:
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.h]: This is the same thing as NSInPredicateOperatorType, except with the left and right expressions swapped.

File: platform/darwin/test/MGLFilterTests.m [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+- (void)tearDown
+{
+    [self.mapView.style removeLayer:layer];
+    [self.mapView.style removeSource:source];
+}
+
+- (NSArray<NSPredicate *> *)predicates
+{
+    NSPredicate *equalPredicate = [NSPredicate predicateWithFormat:@"type == 'neighborhood'"];
```
Comment (1ec5) [inline platform/darwin/test/MGLFilterTests.m]: Mapbox Streets v7 uses the British spelling, `neighbourhood`, for consistency with OpenStreetMap.

File: platform/macos/macos.xcodeproj/project.pbxproj [hunk 16]
```diff
@@ -613,6 +626,7 @@
 			isa = PBXGroup;
 			children = (
 				DA8F257C1D51C5F40010E6B5 /* Layers */,
+				35C5D84B1D6DD75B00E95907 /* MGLFilterTests.m */,
 			);
 			name = Styling;
 			path = ../../ios/test;
```
Comment (1ec5) [inline platform/macos/macos.xcodeproj/project.pbxproj]: This group should correspond to platform/macos/test/, not platform/ios/test/.

File: platform/darwin/src/MGLStyleLayer.h.ejs [hunk 17]
```diff
@@ -35,6 +35,10 @@ typedef NS_ENUM(NSUInteger, MGL<%- camelize(type) %>StyleLayer<%- camelize(prope
 <% } -%>
 @interface MGL<%- camelize(type) %>StyleLayer : MGLBaseStyleLayer <MGLStyleLayer>
 
+<% if (type !== 'background' && type !== 'raster') { -%>
+@property (nonatomic, nullable) NSPredicate *predicate;
```
Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.h.ejs]: Needs documentation.

File: platform/darwin/src/NSComparisonPredicate+MGLAdditions.mm [hunk 18]
```diff
@@ -0,0 +1,76 @@
+#import "NSComparisonPredicate+MGLAdditions.h"
+
+#import "NSPredicate+MGLAdditions.h"
+#import "NSExpression+MGLAdditions.h"
+
+class FilterEvaluator {
+public:
+
+mbgl::style::Filter operator()(NSComparisonPredicate *predicate) {
```
Comment (1ec5) [inline platform/darwin/src/NSComparisonPredicate+MGLAdditions.mm]: This function should be inlined into `-[NSComparisonPredicate(MGLAdditions) mgl_filter]`.

File: platform/darwin/src/NSComparisonPredicate+MGLAdditions.mm [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        case NSMatchesPredicateOperatorType:
+        case NSLikePredicateOperatorType:
+        case NSBeginsWithPredicateOperatorType:
+        case NSEndsWithPredicateOperatorType:
+        case NSCustomSelectorPredicateOperatorType:
+        case NSContainsPredicateOperatorType:
+        case NSBetweenPredicateOperatorType:
+            [NSException raise:@"Operator type not handled"
+                        format:@""];
```
Comment (1ec5) [inline platform/darwin/src/NSComparisonPredicate+MGLAdditions.mm]: For debugging purposes, the format string should say which unsupported operator type is being used here.

File: platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+    
+    std::vector<mbgl::style::Filter>getFilters (NSArray<NSPredicate *> *predicates) {
+        std::vector<mbgl::style::Filter>filters;
+        for (NSPredicate *predicate in predicates) {
+            filters.push_back(predicate.mgl_filter);
+        }
+        return filters;
+    }
+    
+    mbgl::style::Filter operator()(NSCompoundPredicate *predicate) {
```
Comment (1ec5) [inline platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm]: This function should be inlined into `-[NSCompoundPredicate(MGLAdditions) mgl_filter]`.

File: platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm [hunk 21]
```diff
@@ -0,0 +1,51 @@
+#import "NSCompoundPredicate+MGLAdditions.h"
+
+#import "NSPredicate+MGLAdditions.h"
+#import "NSExpression+MGLAdditions.h"
+
+class PredicateEvaluator {
+public:
+    
+    std::vector<mbgl::style::Filter>getFilters (NSArray<NSPredicate *> *predicates) {
```
Comment (1ec5) [inline platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm]: This function should become `-[NSCompoundPredicate(MGLAdditions) mgl_subfilters]`.

File: platform/darwin/src/NSExpression+MGLAdditions.mm [hunk 22]
```diff
@@ -0,0 +1,60 @@
+#import "NSExpression+MGLAdditions.h"
+
+class FilterExpressionEvaluator {
+public:
+    id operator()(mbgl::Value value) {
```
Comment (1ec5) [inline platform/darwin/src/NSExpression+MGLAdditions.mm]: Instead of a single `mbgl::Value` operator with a number of `if…is` clauses, there should be a separate operator for each type you want to convert. So, for example:

```objc
id operator()(bool value) {
    return @(value.get<bool>());
}
```

Comment (1ec5) [inline platform/darwin/src/NSExpression+MGLAdditions.mm]: The implementation of FilterExpressionEvaluator would look identical to PropertyValueEvaluator in MGLFeature.mm. We should make that class accessible to the whole project, then reusing it instead of reinventing the wheel here.

File: platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+    mbgl::style::Filter operator()(NSCompoundPredicate *predicate) {
+        switch (predicate.compoundPredicateType) {
+            case NSNotPredicateType: {
+                auto filter = mbgl::style::NoneFilter();
+                filter.filters = getFilters(predicate.subpredicates);
+                return filter;
+            }
+            case NSAndPredicateType: {
+                auto filter = mbgl::style::AllFilter();
+                filter.filters = getFilters(predicate.subpredicates);
```
Comment (1ec5) [inline platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm]: The NSCompoundPredicate documentation mentions this edge case:

> An AND predicate with no subpredicates evaluates to TRUE.

Does `mbgl::style::AllFilter` have the same behavior?

File: platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+                return filter;
+            }
+            case NSOrPredicateType: {
+                auto filter = mbgl::style::AnyFilter();
+                filter.filters = getFilters(predicate.subpredicates);
```
Comment (1ec5) [inline platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm]: The NSCompountPredicate documentation mentions this edge case:

> An OR predicate with no subpredicates evaluates to FALSE.

Does `mbgl::style::AnyFilter` have the same behavior?

File: platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+        for (NSPredicate *predicate in predicates) {
+            filters.push_back(predicate.mgl_filter);
+        }
+        return filters;
+    }
+    
+    mbgl::style::Filter operator()(NSCompoundPredicate *predicate) {
+        switch (predicate.compoundPredicateType) {
+            case NSNotPredicateType: {
+                auto filter = mbgl::style::NoneFilter();
```
Comment (1ec5) [inline platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm]: This should be `NotFilter`, not `NoneFilter`. Also, for performance reasons, we should have a special case in which an `NSNotPredicateType` predicate containing an `NSInPredicateOperatorType` should use `mbgl::style::NotInFilter`.

File: platform/darwin/test/MGLFilterTests.m [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+{
+    [self.mapView.style removeLayer:layer];
+    [self.mapView.style removeSource:source];
+}
+
+- (NSArray<NSPredicate *> *)predicates
+{
+    NSPredicate *equalPredicate = [NSPredicate predicateWithFormat:@"type == 'neighbourhood'"];
+    NSPredicate *notEqualPredicate = [NSPredicate predicateWithFormat:@"type != 'park'"];
+    NSPredicate *greaterThanPredicate = [NSPredicate predicateWithFormat:@"%K > %@", @"stroke-width", @2.1];
```
Comment (1ec5) [inline platform/darwin/test/MGLFilterTests.m]: Nit: it should also be possible to inline the 2.1 into the format string.

File: platform/darwin/test/MGLFilterTests.m [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+             typePredicate,
+             idPredicate,
+             specialCharsPredicate];
+}
+
+- (void)testAllPredicates
+{
+    for (NSPredicate *predicate in self.predicates) {
+        layer.predicate = predicate;
+        XCTAssertEqualObjects(layer.predicate, predicate);
```
Comment (1ec5) [inline platform/darwin/test/MGLFilterTests.m]: This ensures we can round-trip the predicates, which is a good start. However, the tests don’t verify that for instance `specialCharsPredicate` resolves to the right `mbgl::style::Filter`. What if the special characters get mangled during the conversion but get unmangled during the conversion back to NSPredicate?

File: platform/darwin/src/MGLCircleStyleLayer.h [hunk 28]
```diff
@@ -18,6 +18,16 @@ typedef NS_ENUM(NSUInteger, MGLCircleStyleLayerCirclePitchScale) {
 
 @interface MGLCircleStyleLayer : MGLBaseStyleLayer <MGLStyleLayer>
 
+/**
+ A predicate that corresponds to the layer's filter.
+ 
+ The predicate's left expression must be a string that identifies a feature
+ property, or one of the following special keys:
+    "$type" - identifies the feature by type.
```
Comment (1ec5) [inline platform/darwin/src/MGLCircleStyleLayer.h]: I don’t think this will look like a list in jazzy. We should either insert our own Doxygen or HTML formatting or override this documentation.

File: platform/darwin/src/NSExpression+MGLAdditions.mm [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+{
+    if ([value isKindOfClass:NSString.class]) {
+        return { std::string([(NSString *)value UTF8String]) };
+    } else if ([value isKindOfClass:NSNumber.class]) {
+        NSNumber *number = (NSNumber *)value;
+        if((strcmp([number objCType], @encode(int))) == 0) {
+            return { number.intValue };
+        } else if ((strcmp([number objCType], @encode(double))) == 0) {
+            return { number.doubleValue };
+        } else if ((strcmp([number objCType], @encode(bool))) == 0) {
```
Comment (boundsj) [inline platform/darwin/src/NSExpression+MGLAdditions.mm]: @frederoni  I expect to be able to write something like `[NSPredicate predicateWithFormat:@"%K != %@", @"cluster", [NSNumber numberWithBool:YES]]]` however this is not caught by this implementation (ref: http://stackoverflow.com/questions/23425187/nsnumber-with-bool-is-no-bool-on-64bit). Using `else if ((strcmp([number objCType], @encode(char))) == 0)` worked for me.

Comment (frederoni) [inline platform/darwin/src/NSExpression+MGLAdditions.mm]: Did it raise an exception or returned `mbgl::Value` with incorrect type? Feel free to push the fix.

Comment (boundsj) [inline platform/darwin/src/NSExpression+MGLAdditions.mm]: It triggered the exception since it was not caught in the previous conditions. Fixed here https://github.com/mapbox/mapbox-gl-native/pull/6049/commits/eefe5cb050789078b7fae0f05f13fbc929731240

**Comment by boundsj** [comment]
👍

File: platform/darwin/src/MGLFillStyleLayer.h [hunk 30]
```diff
@@ -13,6 +13,14 @@ typedef NS_ENUM(NSUInteger, MGLFillStyleLayerFillTranslateAnchor) {
 
 @interface MGLFillStyleLayer : MGLBaseStyleLayer <MGLStyleLayer>
 
+/**
+ A predicate that corresponds to the layer's <a href='https://www.mapbox.com/mapbox-gl-style-spec/#types-filter'>filter</a>.
+ 
+ The predicate's left expression must be a string that identifies a feature
+ property, or one of the special keys.
```
Comment (1ec5) [inline platform/darwin/src/MGLFillStyleLayer.h]: Possibly tail work, but we need to document inline which NSExpression operator types we support, as well as the valid values of `$type` and `$id`. Linking to the style specification will suffice for now, but it’s suboptimal because the JavaScript syntax and terminology is very foreign to the NSPredicate-based interface we ended up exposing.

File: platform/darwin/src/NSPredicate+MGLAdditions.mm [hunk 31]
```diff
@@ -0,0 +1,107 @@
...
+        if (predicates.count > 1) {
+            NSCompoundPredicate *predicate = [NSCompoundPredicate andPredicateWithSubpredicates:predicates];
+            return [NSCompoundPredicate notPredicateWithSubpredicate:predicate];
+        } else {
+            return [NSCompoundPredicate notPredicateWithSubpredicate:predicates.firstObject];
+        }
+    }
+    
+    NSPredicate* operator()(mbgl::style::HasFilter filter) {
+        [NSException raise:@"Unsupported filter type"
```
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.mm]: This should be translated as `key != nil`.

File: platform/darwin/src/NSPredicate+MGLAdditions.h (hunk 13 above)
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.h]: → #6244

File: platform/darwin/src/NSPredicate+MGLAdditions.h (hunk 14 above)
Comment (1ec5) [inline platform/darwin/src/NSPredicate+MGLAdditions.h]: → #6244

File: platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm (hunk 23 above)
Comment (1ec5) [inline platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm]: [Yes, it does](https://github.com/mapbox/mapbox-gl-native/blob/6a39cf5aaece81c7a531b12321dd503004cc45b8/include/mbgl/style/filter_evaluator.hpp#L102).

File: platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm (hunk 24 above)
Comment (1ec5) [inline platform/darwin/src/NSCompoundPredicate+MGLAdditions.mm]: [Yes, it does](https://github.com/mapbox/mapbox-gl-native/blob/6a39cf5aaece81c7a531b12321dd503004cc45b8/include/mbgl/style/filter_evaluator.hpp#L93).

File: platform/darwin/src/MGLFillStyleLayer.h (hunk 30 above)
Comment (1ec5) [inline platform/darwin/src/MGLFillStyleLayer.h]: Oh, I’m seeing now that this documentation was originally inline but was removed in response to https://github.com/mapbox/mapbox-gl-native/pull/6049#discussion_r76650575. I meant that we should describe the `$type` and `$id` keys but do so in an HTML definition list.

File: platform/darwin/test/MGLFilterTests.mm [hunk 32]
```diff
@@ -0,0 +1,139 @@
...
+    NSPredicate *greaterThanOrEqualToPredicate = [NSPredicate predicateWithFormat:@"%K >= %@", @"stroke-width", @2.1];
+    NSPredicate *lessThanOrEqualToPredicate = [NSPredicate predicateWithFormat:@"%K <= %@", @"stroke-width", @2.1];
+    NSPredicate *lessThanPredicate = [NSPredicate predicateWithFormat:@"%K < %@", @"stroke-width", @2.1];
+    NSPredicate *inPredicate = [NSPredicate predicateWithFormat:@"type IN %@", @[@"park", @"neighbourhood"]];
+    NSPredicate *notInPredicate = [NSPredicate predicateWithFormat:@"NOT (type IN %@)", @[@"park", @"neighbourhood"]];
+    NSPredicate *inNotInPredicate = [NSPredicate predicateWithFormat:@"type IN %@ AND NOT (type IN %@)", @[@"park"], @[@"neighbourhood", @"test"]];
+    NSPredicate *typePredicate = [NSPredicate predicateWithFormat:@"%K == %@", @"$type", @"Feature"];
+    NSPredicate *idPredicate = [NSPredicate predicateWithFormat:@"%K == %@", @"$id", @"1234123"];
+    NSPredicate *specialCharsPredicate = [NSPredicate predicateWithFormat:@"%K == %@", @"ty-’pè", @"sŒm-ethįng"];
+    NSPredicate *booleanPredicate = [NSPredicate predicateWithFormat:@"%K != %@", @"cluster", [NSNumber numberWithBool:YES]];
```
Comment (1ec5) [inline platform/darwin/test/MGLFilterTests.mm]: This can be inlined as `%K != YES`.
